src/server/daemons/GenerateRecommendationDaemon.java

```diff
@@ -0,0 +1,147 @@
+package server.daemons;
+
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import data.proxy.PreferenceCorrelationGraph;
+import data.structure.Preference;
+import data.structure.PreferenceCategory;
+import data.structure.PreferenceCorrelation;
+import data.structure.Recommendation;
+import data.structure.Tuple2;
+import data.structure.UserProfile;
+
+/**
+ * A service (not really a daemon) that generates a recommendation for an item that does not match one of the user's existing preferences.  It uses
+ *    correlation scores to determine which preference has the highest total sum of correlations with the user's existing preferences.  It is attempting
+ *    to be scalable by handling preferences in batches, given that there could be a very large number of preferences in total, even within a given category.
+ *    Currently recalculates all preference scores from scratch every time it is called, but will soon implement logic to allow for simply updating correlation
+ *    scores based on update events.
+ * 
+ * @author benjaminposerow
+ *
+ */
+public class GenerateRecommendationDaemon {
+	private int batchSize;
+	
+	/**
+	 * 
+	 * @param batchSize  The number of preferences that are processed at once in the calculation of preference correlation scores between the universe of
+	 *    existing preferences in a category and user preferences.  Doing these calculations in batches allows for greater opportunities for parallelism in
+	 *    the determination of the best recommendation for a user amongst a potentially very large number of preferences.  It also potentially could reduce
+	 *    demand for memory, because you don't need to load all preferences to do the calculation and you could potentially serialize the correlation scores
+	 *    to DDB for each batch before processing the next batch.
+	 */
+	public GenerateRecommendationDaemon(int batchSize) {
+		this.batchSize = batchSize;
+	}
+	
+	Map<Preference, Double> getBatchScores(UserProfile user,
+			List<Preference> preferenceBatch,
+			PreferenceCategory preferenceCategory) {
+		Set<Preference> userPreferences = user.getPreferencesForCategory(preferenceCategory);
+		
+		// Find those preferences that don't already belong to the user
+		Set<Preference> candidatePreferences = getNonUserPreferences(preferenceBatch, userPreferences);
+		
+		// Take the sum of the correlation ratios between each pair of user preferences and other candidate preferences.  Sum the
+		//   correlation scores across all user preferences.  Return a map of candidate preference to the sum of the correlation scores for
+		//   that preference.
+		return candidatePreferences.stream().map(pref -> new Tuple2<Preference, Double>(pref, 
```

Comment:
I extracted into a separate method.  But I'm not sure that breaking up this expression makes sense.  It's a more functional style, but I feel that chaining together multiple such method calls i quite declarative and almost self documenting, and I also feel all of these operations belong together (they are working together to perform one operation).  I could theoretically assign intermediate results to variables or extract into separate methods but I'm not sure such refactoring will add to clarity rather than hinder it.